res/controllers/Denon-Prime-4-scripts.js

```diff
@@ -0,0 +1,884 @@
+        Prime4.effectBank[1].unshift();
+    }
+};
+
+//========== PERFORMANCE PADS ==========//
+
+// Access the appropriate mode-select pad without remembering MIDI values
+Prime4.padMode = {
+    HOTCUE: 0x0B,
+    LOOP: 0x0C,
+    ROLL: 0x0D,
+    SLICER: 0x0E,
+};
+
+Prime4.PadSection = function(deck, offset) {
+    components.ComponentContainer.call(this);
+
+    // Create component containers for each pad mode
+    const modes = new components.ComponentContainer({
+        "hotcue": new Prime4.hotcueMode(deck, offset),
+        "loop": new Prime4.loopMode(deck, offset),
+        //TODO: "autoloop": new Prime4.autoloopMode(deck, offset),
+        "roll": new Prime4.rollMode(deck, offset),
+        "sampler": new Prime4.samplerMode(deck, offset),
+    });
+
+    const controlToPadMode = control => {
+        let mode;
+        switch (control) {
+        case Prime4.padMode.HOTCUE:
+            mode = modes.hotcue;
+            break;
+        case Prime4.padMode.LOOP:
+            mode = modes.loop;
+            /*
+            // TODO: Implement autoloop mode
+            if (Prime4.PadSection.currentMode === modes.loop) {
+                mode = modes.autoloop;
+            } else {
+                mode = modes.loop;
+            }
+            */
+            break;
+        case Prime4.padMode.ROLL:
+            mode = modes.roll;
+            break;
+        case Prime4.padMode.SLICER:
+            mode = modes.sampler;
+            break;
+        }
+
+        return mode;
+    };
+
+    this.offset = offset;
+
+    // Function for switching between pad modes
+    this.setPadMode = function(control) {
+        const newMode = controlToPadMode(control);
+
+        // Exit early if requested mode is already active or unavailable
+        if (newMode === this.currentMode || newMode === undefined) {
+            return;
+        }
+
+        // Disable LED of current mode button
+        if (this.currentMode) {
+            midi.sendShortMsg(0x94 + this.offset, this.currentMode.ledControl, this.currentMode.colourOff);
+
+            // Disconnect pads from current mode
+            this.currentMode.forEachComponent(function(component) {
+                component.disconnect();
+                component.outConnect = false;
+            });
+        }
+
+        // Enable LED of new mode to switch to
+        midi.sendShortMsg(0x94 + this.offset, newMode.ledControl, newMode.colourOn);
+
+        // Connect pads to new mode
+        newMode.forEachComponent(function(component) {
+            component.outConnect = true;
+            component.connect();
+            component.trigger();
+        });
+
+        this.currentMode = newMode;
+    };
+
+    // Start in Hotcue mode
+    this.setPadMode(Prime4.padMode.HOTCUE);
+    midi.sendShortMsg(0x94 + offset, 0x0C, Prime4.rgbCode.whiteDark);
+    midi.sendShortMsg(0x94 + offset, 0x0D, Prime4.rgbCode.whiteDark);
+    midi.sendShortMsg(0x94 + offset, 0x0E, Prime4.rgbCode.whiteDark);
```

Comment:
Instead of manually sending midi messages, we'll use a similar approach to the JogWheelLED component: 

Add this somewhere in the constructor:
```javascript
this.padModeSelectLeds = components.Component({
    trigger: function() {
        modes.forEach(mode => {
            midi.sendShortMsg(0x94 + offset, mode.ledControl, theContainer.currentMode === mode ? mode.colourOn : mode.colourOff);
        });
    },
});
```
don't forget to add `const theContainer = this;` to somewhere in the constructor so we can access it in the component (since it overwrites `this`).

Then, you'll simply need to trigger that LED output when changing the padmode, if you change the deck, the components `trigger` method will called automatically. 

```suggestion
        if (this.currentMode) {
            // Disconnect pads from current mode
            this.currentMode.forEachComponent(function(component) {
                component.disconnect();
                component.outConnect = false;
            });
        }
        // Connect pads to new mode
        newMode.forEachComponent(function(component) {
            component.outConnect = true;
            component.connect();
            component.trigger();
        });

        this.currentMode = newMode;
        
        theContainer.padModeSelectLeds.trigger();
    };

    // Start in Hotcue mode
    this.setPadMode(Prime4.padMode.HOTCUE);
```

Reply:
Currently getting this error: 
```
Uncaught exception: file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:764: TypeError: Cannot call method 'trigger' of undefined
Backtrace: @file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:764
@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:768
@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:512
@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:267
```
I've pushed the code in its current state for inspection